R/functions.R

```diff
@@ -0,0 +1,46 @@
+
+#' @export
+#' @rdname properties
+lightness <- function (shades, values = NULL) UseMethod("lightness")
+
+#' @export
+#' @rdname properties
+lightness.function <- function(shades, values = NULL) {
+  orig_call <- match.call(expand.dots = FALSE)
+  orig_call[[1]] <- quote(lightness)
+  palette <- shades
+  palette_q <- substitute(shades)
+  orig_env <- parent.frame()
+  
+  wrapper <- function (...) {
+    inner_call <- match.call(expand.dots = FALSE)
+    inner_call[[1]] <- palette_q
+    result <- eval(inner_call, envir = orig_env)
+    orig_call$shades <- result
+    eval(orig_call)
+  }
+  
+  formals(wrapper) <- formals(palette)
+  class(wrapper) <- class(palette)
+   
+  return(wrapper)
+}
+
+#' @export
+#' @rdname properties
+lightness.ggproto_method <- lightness.function
+
+#' @export
+#' @rdname properties
+lightness.Scale <- function (shades, values = NULL) {
+  orig_call <- match.call(expand.dots = FALSE)
+  orig_call[[1]] <- quote(lightness)
+  
+  new_scale <- ggplot2::ggproto(NULL, shades,
+        palette = function (self, n) {
+          res <- ggplot2::ggproto_parent(shades, self)$palette(n)
+          lightness(res, values)
+        })
+
+  new_scale
+}
```

Comment:
You clearly know a lot more about `ggplot2`'s internals than me! This final method seems to not be a straight application to a function, instead requiring a prototype object to be created – could you clarify the use-case for this version, please?

Reply:
This is what allows the user to do e.g.

    lightness(scale_fill_continuous(), 0.5)

to tweak ggplot colours. I learnt about the internals by figuring this out! A simpler version would just directly replace the `palette` member of the scales object, that seems more intrusive but perhaps it would be fine. `?ggplot2::Scale` doesn't really offer any guarantees about what scales objects contain, but I doubt the palette member is going away soon.